modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js

```diff
@@ -159,57 +120,81 @@ function DataSetDisplay({
 			delta,
 			pageNumber,
 			sorting
-		)
-			.then(updateDataSetItems)
-			.then(() => {
-				const {message, showSuccessNotification} = successNotification;
-
-				if (showSuccessNotification) {
-					openToast({
-						message:
-							message ||
-							Liferay.Language.get('table-data-updated'),
-						type: 'success',
-					});
-				}
+		).catch((error) => {
+			logError(error);
+			openToast({
+				message: Liferay.Language.get('unexpected-error'),
+				type: 'danger',
+			});
 
-				setLoading(false);
-				Liferay.fire(DATASET_DISPLAY_UPDATED, {id});
-			})
-			.catch((error) => {
-				logError(error);
-				setLoading(false);
+			throw error;
+		});
+	}, [apiURL, currentURL, delta, filters, pageNumber, searchParam, sorting]);
+
+	const requestComponent = useCallback(() => {
+		if (
+			!CurrentViewComponent &&
+			(contentRendererModuleURL || contentRenderer)
+		) {
+			return (contentRenderer
+				? getViewContentRenderer(contentRenderer)
+				: getJsModule(contentRendererModuleURL)
+			).catch((error) => {
+				logError(
+					`Requested module: ${contentRendererModuleURL} not available`,
+					error
+				);
 				openToast({
 					message: Liferay.Language.get('unexpected-error'),
 					type: 'danger',
 				});
+
+				throw error;
 			});
+		}
+
+		return Promise.resolve(CurrentViewComponent);
+	}, [contentRenderer, contentRendererModuleURL, CurrentViewComponent]);
+
+	const isMounted = useIsMounted();
+
+	function updateDataSetItems(dataSetData) {
+		setTotal(dataSetData.totalCount);
+		updateItems(dataSetData.items);
 	}
 
-	const {apiURL} = useContext(AppContext);
 	useEffect(() => {
-		if (apiURL) {
-			getData(
-				apiURL,
-				currentURL,
-				filters.filter((filter) => filter.value),
-				searchParam,
-				delta,
-				pageNumber,
-				sorting,
-				false
-			);
-		}
-		// eslint-disable-next-line react-hooks/exhaustive-deps
+		let resolved = false;
```

Comment:
My only thought here is that you could conceivably have multiple effects running at once (ie. I click on a second page before the first one has finished loading).

So, they'll both have their own closure-private copy of the `resolved` variable and will happily do their own thing without knowledge that the other one is running. Combine this with the fact that you can't cancel a promise and we have to do some external book-keeping to make sure only one of these things is happening at a time.

This suggests to me that instead of a local variable, we should be using a `useRef` ref which will be shared by any concurrent effects that might be running. And you could store the currently pending promise in it as a way of achieving cancellation semantics.

For example:

```js
const pendingPromise = useRef(null);

useEffect() => {
    const promise = Promise.race([
        delay(200).then(() => {
            if (pendingPromise.current === promise) {
                // At this point, you know you haven't been "canceled"
                // and can get on with your business,    
            }
        }),
        // etc... with a similar check...
    ]);

    pendingPromise.current = promise;
}, [...]);
```

Reply:
![principio](https://user-images.githubusercontent.com/156388/91480384-46848c00-e879-11ea-9cff-aa3122ffecef.jpg)

Reply:
I've applied your suggestion and force pushed. Thank you.

I had to do https://github.com/liferay-frontend/liferay-portal/pull/318/commits/575211ca67264206e3f138d269137856b1b66108#diff-a50a96ebe561378b0b93b9dcca5b10feR178 otherwise the delay callback would come in late and show the loading state after wall was done.

Reply:
Makes sense. Good stuff.